jlm/tooling/Command.cpp

```diff
@@ -338,21 +335,62 @@ JlmOptCommand::Run() const
   statisticsCollector.PrintStatistics();
 }
 
-std::unique_ptr<::llvm::Module>
-JlmOptCommand::ParseLlvmIrFile(const util::filepath & llvmIrFile, ::llvm::LLVMContext & llvmContext)
-    const
+std::unique_ptr<llvm::RvsdgModule>
+JlmOptCommand::ParseInputFile(
+    const util::filepath & inputFile,
+    const JlmOptCommandLineOptions::InputFormat & inputFormat,
+    util::StatisticsCollector & statisticsCollector) const
 {
-  ::llvm::SMDiagnostic diagnostic;
-  if (auto module = ::llvm::parseIRFile(llvmIrFile.to_str(), diagnostic, llvmContext))
+  auto parseLlvmIrFile =
+      [=](const util::filepath & llvmIrFile,
+          util::StatisticsCollector & statisticsCollector) -> std::unique_ptr<llvm::RvsdgModule>
   {
-    return module;
-  }
+    ::llvm::LLVMContext llvmContext;
+    ::llvm::SMDiagnostic diagnostic;
+    auto llvmModule = ::llvm::parseIRFile(llvmIrFile.to_str(), diagnostic, llvmContext);
 
-  std::string errors;
-  ::llvm::raw_string_ostream os(errors);
-  diagnostic.print(ProgramName_.c_str(), os);
-  throw util::error(errors);
-}
+    if (llvmModule == nullptr)
+    {
+      std::string errors;
+      ::llvm::raw_string_ostream os(errors);
+      diagnostic.print(ProgramName_.c_str(), os);
+      throw util::error(errors);
+    }
+
+    auto interProceduralGraphModule = llvm::ConvertLlvmModule(*llvmModule);
+
+    // Dispose of Llvm module. It is no longer needed.
+    llvmModule.reset();
+
+    auto rvsdgModule =
+        llvm::ConvertInterProceduralGraphModule(*interProceduralGraphModule, statisticsCollector);
+
+    return rvsdgModule;
+  };
+
+  auto parseMlirIrFile =
+      [](const util::filepath & mlirIrFile,
+         util::StatisticsCollector & statisticsCollector) -> std::unique_ptr<llvm::RvsdgModule>
+  {
+#ifdef ENABLE_MLIR
+    jlm::mlir::MlirToJlmConverter rvsdggen;
+    return rvsdggen.ReadAndConvertMlir(mlirIrFile);
+#else
+    JLM_UNREACHABLE(
+        "This version of jlm-opt has not been compiled with support for the MLIR backend\n");
+#endif
+  };
+
+  static std::unordered_map<
+      JlmOptCommandLineOptions::InputFormat,
+      std::function<
+          std::unique_ptr<llvm::RvsdgModule>(const util::filepath &, util::StatisticsCollector &)>>
+      parsers({ { tooling::JlmOptCommandLineOptions::InputFormat::Llvm, parseLlvmIrFile },
+                { tooling::JlmOptCommandLineOptions::InputFormat::Mlir, parseMlirIrFile } });
+
+  JLM_ASSERT(parsers.find(inputFormat) != parsers.end());
+  return parsers[inputFormat](inputFile, statisticsCollector);
```

Comment:
This is just two different types of parsers. I think an if-then-else statement would be more readable:

```
if (inputFormat == tooling::JlmOptCommandLineOptions::InputFormat::Llvm)
{
  return parseLlvmIrFile(...);
}
else if (inputFormat ....)
{
  return parseMlirIrFile(...);
}
else
{
  JLM_UNREACHABLE("Unhandled input format.");
}
```

Reply:
It's basically a copy of how the OutputFormat is handled, which before this PR also had 2 types (xml and llvm)
https://github.com/phate/jlm/blob/fef06f81a60c0b6a01654ec1b89a041167754fa3/jlm/tooling/Command.cpp#L398
So this is consistent with existing code.

But I can of course change it.
Would you in that case want both places to be changed to if/else?

Reply:
Changed to if-then-else for both input and output format.